[GStreamer][WebRTC] DTMF support

philn opened the pull request:
#### 0894e1f17293001ef1c06bf569b9be49a8d58d03
<pre>
[GStreamer][WebRTC] DTMF support
<a href="https://bugs.webkit.org/show_bug.cgi?id=293551">https://bugs.webkit.org/show_bug.cgi?id=293551</a>

Reviewed by Xabier Rodriguez-Calvar.

Our outgoing audio source now includes a dtmf source that can be controlled using the
GStreamerDTMFSenderBackend. Local audio playback of DTMF is handled using a new pipeline, within the
sender backend. When receiving DTMF payloads, webrtcbin currently rejects those buffers, due to:

<a href="https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/4458">https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/4458</a>

This isn&apos;t a critical issue though because the spec doesn&apos;t mandate actual decoding of DTMF
payloads.

Canonical link: <a href="https://commits.webkit.org/295749@main">https://commits.webkit.org/295749@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e54bd2d3b241c9926f523874006996a4437a1f44

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106045 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16188 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111242 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56641 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108084 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26450 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34297 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/80553 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109049 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20889 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95704 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60876 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13802 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56079 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90261 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13837 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114097 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33183 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/24490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/89635 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33547 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91932 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89324 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22762 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34189 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/12002 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28753 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33108 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/38520 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32854 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36204 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34452 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3f7424c24ae1877b6537bc036e8d2290a2f7585c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45992

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3b7f16fa4f14e087788c7df47bbac25c38ff46ca. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45992

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c5d3d1f0ed4d99d3550a116c9a7902ede5dd8cb3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45992

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/51d049308b2e178790999d3fd69a8e205c9d414e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45992

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/edb3a83547c117d1cab9318a65d5131c70488640. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45992

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c545d58e1a7004e4ee7c00c3466c2734191cd966. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45992

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/12973ddefb4f082c96394bc448dd2ab5fe70f82c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45992

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b3e431535a24929b3fb3da55bbd42c410b448a4a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45992

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerDTMFSenderBackend.cpp:
```diff
@@ -20,60 +20,197 @@
+
+    if (tone == ',') {
+        GST_DEBUG_OBJECT(element.get(), "Playing silence for 2 seconds");
+        sleep(Seconds::fromMilliseconds(2000));
```
Why not using the _s operator just as 2_s?

philn replied:
Because I forgot we had this operator :)

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerDTMFSenderBackend.cpp:
```diff
@@ -20,60 +20,197 @@
+    data->pad = WTFMove(pad);
+    data->self = this;
+
+    g_timeout_add_full(G_PRIORITY_DEFAULT, duration, reinterpret_cast<GSourceFunc>(+[](gpointer userData) -> gboolean {
```
Why can't you just use webkit function to run on main thread?

philn replied:
That's what came to me first, but yes, a one shot timer should also work.

philn committed:
[GStreamer][WebRTC] DTMF support
https://bugs.webkit.org/show_bug.cgi?id=293551

Reviewed by Xabier Rodriguez-Calvar.

Our outgoing audio source now includes a dtmf source that can be controlled using the
GStreamerDTMFSenderBackend. Local audio playback of DTMF is handled using a new pipeline, within the
sender backend. When receiving DTMF payloads, webrtcbin currently rejects those buffers, due to:

https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/4458

This isn't a critical issue though because the spec doesn't mandate actual decoding of DTMF
payloads.

Canonical link: https://commits.webkit.org/295749@main
